core/api/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+        -> Result<SpendableOutput, DecodeError>;
+
+    /// Decode `Input` compatible with this module, after the module key prefix was already decoded
+    fn decode_input(&self, r: &mut dyn io::Read) -> Result<Input, DecodeError>;
+
+    /// Decode `Output` compatible with this module, after the module key prefix was already decoded
+    fn decode_output(&self, r: &mut dyn io::Read) -> Result<Output, DecodeError>;
+
+    /// Decode `PendingOutput` compatible with this module, after the module key prefix was already decoded
+    fn decode_pending_output(&self, r: &mut dyn io::Read) -> Result<PendingOutput, DecodeError>;
+
+    /// Decode `OutputOutcome` compatible with this module, after the module key prefix was already decoded
+    fn decode_output_outcome(&self, r: &mut dyn io::Read) -> Result<OutputOutcome, DecodeError>;
+}
+
+/// Something that can be an [`Input`] in a [`Transaction`]
+///
+/// General purpose code should use [`Input`] instead
+pub trait ModuleInput: DynEncodable {
+    fn module_key(&self) -> ModuleKey;
+    fn amount(&self) -> Amount;
+    fn clone(&self) -> Input;
+}
+
+def_module_type_newtype! {
+    /// An owned, immutable input to a [`Transaction`]
+    Input(Box<ModuleInput>)
+}
+impl_module_type_newtype_encode! {
+    Input, decode_input
+}
+impl_module_type_newtype_clone!(Input);
+
+/// Something that can be an [`Output`] in a [`Transaction`]
+///
+/// General purpose code should use [`Output`] instead
+pub trait ModuleOutput: DynEncodable {
+    fn module_key(&self) -> ModuleKey;
+    fn amount(&self) -> Amount;
+
+    fn clone(&self) -> Output;
+}
+
+def_module_type_newtype! {
+    /// An owned, immutable output of a [`Transaction`]
+    Output(Box<ModuleOutput>)
+}
+impl_module_type_newtype_encode! {
+    Output, decode_output
+}
+impl_module_type_newtype_clone!(Output);
+
+/// A spendable output - tracked and persisted by the client
+///
+/// Created by generating transaction [`Output`], spendable
+/// by converting to [`Input`].
+pub trait ModuleSpendableOutput: DynEncodable {
+    /// Module key
+    fn module_key(&self) -> ModuleKey;
+    fn amount(&self) -> Amount;
+    fn clone(&self) -> SpendableOutput;
+
+    /// Prepare [`Input`] spending thish output in a transaction, and a key used to sign the [`Transaction`]
+    fn to_input(&self) -> (Input, bitcoin::KeyPair);
+
+    fn key(&self) -> String;
+}
+
+def_module_type_newtype! {
+    /// An owned, immutable output of a [`Transaction`] after it was finalized (so it's spendable)
+    SpendableOutput(Box<ModuleSpendableOutput>)
+}
+impl_module_type_newtype_encode! {
+    SpendableOutput, decode_spendable_output
+}
+impl_module_type_newtype_clone!(SpendableOutput);
+
+pub enum FinalizationError {
+    SomethingWentWrong,
+}
+
+/// A pending output - tracked and persisted by the client
+///
+/// Created by generating transaction [`Output`], spendable
+/// by converting to [`Input`].
+pub trait ModulePendingOutput: DynEncodable {
```

Comment:
That's a good and important question. My naive intuition was to let the client keep track of things, so each `struct` becomes a bit like a state in a state machine, where module-provided logic is responsible for state transformations, and generic code responsible for keeping track of them. Possibly this would allow client-side modules to be (more/completely) stateless themselves. Maybe we wouldn't even need to pass any sort of database handles to them. 

Limitations of this approach that I see is that each input/output is individually abstract, so it's not possible to check the state of all pending mint outputs at the same time, etc. This could be modeled in some other way.

Hybrid approaches are also possible. Clients could keep track of things, register them in a module on initialization, while the modules are still free to run their own threads and do any internal tracking.